dbms/src/DataStreams/WindowBlockInputStream.h

```diff
@@ -171,6 +165,48 @@ struct WindowTransformAction
 
     void appendInfo(FmtBuffer & buffer) const;
 
+private:
+    void stepToFrameStartWithOffsetBoundry();
+    void stepToFrameEndWithOffsetBoundary();
+
+    // Used for calculating the frame start for rows frame type
+    RowNumber stepToFrameStartForRows();
+    // Used for calculating the frame end for rows frame type
+    std::tuple<RowNumber, bool> stepToFrameEndForRows();
+
+    // Used for calculating the frame start for range frame type
+    RowNumber stepToFrameStartForRange();
+    // Used for calculating the frame end for range frame type
+    std::tuple<RowNumber, bool> stepToFrameEndForRange();
+
+    template <bool is_desc>
+    RowNumber stepToFrameStartForRange();
```

Comment:
Maybe rename it to `stepToFrameStartForRangeFrameOrderCase`

Reply:
> Maybe rename it to `stepToFrameStartForRangeFrameOrderCase`

okk